### PR TITLE
feat: torch 1.9.0 + various version upgrades [DET-5547]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description
Versions upgraded:
torch: 1.7.1 -> 1.9.0
torchvision: 0.8.2 -> 0.10.0
- to support PyTorch profiler and CUDA 11
horovod: 0.22.0 -> 0.22.1
- fixes for sparse gradients in torch 1.8.1+
tensorflow: 2.4.1 -> 2.4.2
- various security fixes
pytorch_lightning: 1.2.0 -> 1.3.5
- upgrade pyyaml to 5.4.1 to fix security error preventing images from publishing on Nvidia registry

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: docs/faq.txt [hunk 1]
```diff
@@ -214,10 +214,10 @@ container image that has been configured for that experiment. Determined
 provides prebuilt Docker images that include TensorFlow 1.15, 2.4, and
 2.5, respectively:
 
--  ``determinedai/environments:cuda-10.2-pytorch-1.7-tf-1.15-gpu-0.16.0``
+-  ``determinedai/environments:cuda-10.2-pytorch-1.7-tf-1.15-gpu-0.16.1``
    (default)
```
Comment (ioga) [inline docs/faq.txt:218]: I believe the second one is now the default. can we reorder them and update this accordingly?

**Review by ioga — COMMENTED** [review]
could we run the upstream branch tests on this, just to make sure the CI build the stuff "right"?

File: docs/faq.txt (hunk 1 above)
Comment (azhou-determined) [reply docs/faq.txt]: nice catch. done

**Comment by ioga** [comment]
ah, we also need to update `pytorch-lightning==1.3.5` in `docs/requirements.txt` so they all are in order.

**Comment by azhou-determined** [comment]
> could we run the upstream branch tests on this, just to make sure the CI build the stuff "right"?

enabled GPU e2e tests. they're running here: https://app.circleci.com/pipelines/github/determined-ai/determined?branch=torch-1.9-and-various-upgrades

**Review by ioga — APPROVED** [review]
